src/pal/src/exception/seh-unwind.cpp

```diff
@@ -372,12 +372,14 @@ static volatile size_t s_allocatedContextsBitmap = 0;
 Parameters:
     exceptionRecord - output pointer to the allocated exception record
     contextRecord - output pointer to the allocated context record
+    allocationProhibited - input flag to avoid memory allocation in critical situations
 --*/
 VOID
-AllocateExceptionRecords(EXCEPTION_RECORD** exceptionRecord, CONTEXT** contextRecord)
+AllocateExceptionRecords(EXCEPTION_RECORD** exceptionRecord, CONTEXT** contextRecord, BOOL allocationProhibited)
 {
     ExceptionRecords* records;
-    if (posix_memalign((void**)&records, alignof(ExceptionRecords), sizeof(ExceptionRecords)) != 0)
+    if (allocationProhibited ||
+	(posix_memalign((void**)&records, alignof(ExceptionRecords), sizeof(ExceptionRecords)) != 0) )
```

Comment:
Do we need to change the implementation below to keep re-trying instead of aborting when we do not have a context available?

Reply:
The problem is that we have a limited number of the preallocated contexts (64 for 64 bit processors, 32 for 32 bit ones). That means that if more than that number of threads were handling hardware exceptions at the same time, we would abort the process. In fact, even with less threads, we can exhaust this storage easily if a hardware exception happens while we are handling another one and then another etc. We need to keep the previous exception until the final one is handled.
So it seems that the solution will need to be more involved than using these fallback contexts which were really meant just to be a bit more resilient in out of memory cases.

Reply:
@jkotas retrying would not help, the contexts are alive during the whole lifetime of the PAL_SEHException, which means until the exception is fully handled.

Reply:
I actually wonder if using mmap to allocate memory in the signal handler would be safe. While the mmap is not listed on the async signal safe functions list, I don't see how it could cause a problem. The problems with other functions are based on the fact that the sigsegv can happen inside of such functions and that such functions are not reentrant. But mmap is just a syscall and so no sigsegv can occur while executing code inside of it, thus it should be safe to call from our sigsegv handler.
So maybe we could create a custom allocator based on mmap and use it here.

Reply:
@janvorli Could you explain why preallocated context count is so tiny? For atomic free slot searching? Why not use N bitmaps?

I can add check for recursive handler and using preallocated context only if previous signal handler is not completed. But it ruined main idea do not use unsafe functions in signal handler.

Reply:
Yes, the reason why it is so tiny is so that we can allocate the slot atomically. That's why the count is bitness specific. 
We really need a right fix that works in all cases, so checks for recursive handler would not be the right solution (moreover, you would not be able to track the recursiveness at this level, since we never return back to the signal handler if the exception is processed by the runtime.
I can see a several possible ways we could solve the issue:
1. Use mmap as a base for a custom allocator for the contexts. Either for all or just for the ones allocated for the signal handler. As I've said before, I think there should be no problem with using mmap in the handler even though it is not listed as async signal safe.
2. Refactor the common_signal_handler, SEHProcessException and related code to not to allocate the records until we know that the failing code was executing in code under our control - that means that g_safeExceptionCheckFunction or CatchHardwareExceptionHolder::IsEnabled()returned true. In such case, it is safe to use the existing allocator. That will complicate a bit sharing the SEHProcessException implementation for OSX and other Unixes, since the records are generated differently for OSX and the others and we would need to do the conversion between the platform native context and the 
3. Initialize the PAL_SEHException with records allocated on stack and add a method to the PAL_SEHException that would update the records to heap allocated ones (allocate the memory and copy the original records there). This method would be called from SEHProcessException before the call to the g_hardwareExceptionHandler and PAL_ThrowExceptionFromContext where we know it is safe to allocate. We would add a bool flag into the PAL_SEHException indicating whether the records are on stack or on heap to make sure the destructor does not try to free them if they were still on stack. That would also make it transparent to the OSX case where the records will always be allocated from the heap, since on OSX, we don't use signals for handling hardware exceptions.

I think that we should go with the way 3 - we could initiate the exception with the signalContextRecord (and change the code in the common_signal_handler to extract the context directly into it). Since we already had a memcpy that copied the context to this structure, we would not be adding an additional copy of the large record (just move the place where we do the copy), which is good.
Also, the change will be quite localized and won't do anything that's not safe.

Reply:
1. I agree. Only performance problem may occur. Because mmap is a system call moreover it is serialized on process system call.

2 - 3.  Signal handler is not a safe place for glibc memory allocation. OK some cases are safe for runtime. But all are not safe for other (native) threads.  Original issue #16338 catches segfault not in memory allocation but when memory is fried. In this case we can't use glibc allocation at all.  Also delayed allocation is useless because it is only optimization for signals generated from not managed code.

So I think we have to use mmap for allocation exception context in signal handler.

Reply:
There is no way to avoid allocations in the signal handler. The whole hardware exception handling is executed from the handler. That means that once we start handling the exception, there is no limit in what we call and execute and we never return from the signal handler.
But what is important is to not to start calling async signal unsafe stuff until we know that the code that was executing was managed code or special helpers in our runtime. Once we are sure it was the case, there is no issue with calling anything on the handler, since we know we won't be reentering any platform functions.
That's why I believe we should use the way #3. The way #1 would require writing an allocator on top of mmap so that we don't waste memory (especially on arm64 where page size is 64 kB on some Linux distros). That would make it considerably more complex than the way #3.